### ISSUE TITLE
Whether new releases of r4ss should maintain backwards compatibility

### ISSUE THREAD

**Issue by ChristineStawitz-NOAA** [r4ss/r4ss#306]
@CorinneBassin-NOAA and I are working on defining recommended workflows for the NOAA Fisheries Toolbox, and we discussed using github's release feature to simplify branches in code flow. I think breaking backwards compatibility for past SS versions in future releases of r4ss has the potential to simplify the code and improve stability and testability. Although the way the current functions like `read_dat` are modularized is the best way to have a fully-backwards-compatible r4ss, it still seems that many issues are caused when adding functionality for SS 3.30 models breaks functionality for SS 3.24, etc. Furthermore the test suite is much larger when the latest r4ss version needs to work across SS versions.

What I am envisioning is that specific r4ss releases work for some versions but not others; for example, r4ss v x.00 is fully tested and works for SS 2.0 models but not SS 3.30 models, but version x+1.00 of r4ss works for ss 3.30 models only and not SS 2.0 models.

I suggest this because I think it might make your lives easier, but if it seems to make things harder instead (i.e. because you plan to add new features for older SS versions) please let me know. I am hoping to start a conversation, not be prescriptive about changes.

**Comment by iantaylor-NOAA** [comment]
Thanks @ChristineStawitz-NOAA and @CorinneBassin-NOAA for starting this discussion. I agree that maintaining backwards compatibility adds challenges and makes the code more complex and harder to test. However, there are also a variety of areas where it's been really useful.

Two general cases where maintaining compatibility of r4ss with both 3.24 and 3.30 are the following.

**Comparing old and new SS models for the same stock**
The Pacific Fishery Management Council Terms of Reference for groundfish stock assessments (https://www.pcouncil.org/wp-content/uploads/2019/04/Stock_Assessment_ToR_REVISED_2019-20_APR2019_Final-2.pdf) require a description of the bridging from the most recent previous assessment model to the new base model. Plots comparing biomass time series using the `SSplotComparisons` function has been a useful part of this description. The 2017 assessments were conducted in SS version 3.30, but many assessments conducted in 2015 and before have not yet been replaced by full assessments, and would benefit from r4ss compatibility when that happens. I suspect a similar process is in place for other groups who use SS. However, even without compatibility, it's not that hard to extract a biomass time series from an old version of a Report file and add it to the input to `SSplotComparisons`. 

As another example, Neal Klaer, Rick Methot and Andre Punt are involved in an effort to summarize best practices for stock assessments for which Neal Klaer contributed `readdat` functions so that he could aggregate information from a large set of SS models spanning many versions.

**Assessments that continue to stick with older SS versions**
Many assessments continue to be conducted in flavors of SS version 3.24. In some cases these are things like catch-only updates, where updating the software is not part of the process. In other cases it might be just to save the effort to make the switch. Some assessment authors might prefer to use the exact r4ss release that was first associated with this model, in which case backwards compatibility isn't an issue. However, in other cases, there may be new r4ss features or bug fixes that would be useful to apply to the older model version. 


At the moment, dropping compatibility with SS version 3.24 doesn't strike me as enough benefit to outweigh the effort and inconvenience it might cause, but I see no need to maintain compatibility of current versions of r4ss with any SS version older than 3.24. An old figure illustrating the timeline of SS releases during the 2010-2013 period is below, indicating that 3.24 started in 2012, so anything prior could be considered ancient history.  Thus, I support the suggestion to choose a range of versions to maintain compatibility with. The question would then become how broad a range to maintain and when to revise it.

![image](https://user-images.githubusercontent.com/4992918/58671621-1e198500-82f8-11e9-9c70-2a0b72484188.png)

**Comment by kellijohnson-NOAA** [comment]
Great topic. I struggled with this regarding ss3sim as well. @k-doering-NOAA and I decided to not maintain backwards compatibility with 3.24 as we transition to 3.30 because we believe that new simulations should only be based on the best available software, i.e., the most up to date version of SS. Here, with r4ss, I think it is different because like @iantaylor-NOAA mentioned, there are studies that use r4ss to read in old results. Honestly, I think maintaining backwards compatibility will facilitate better research because it allows for easier comparison of more assessments than if compatibility were to be lost. 
I do think that there are ways in which the compatibility could be streamlined and I am happy to help brainstorm best practices here. 
In short, I think that the NOAA Toolbox might want to think about advocating for backwards compatibility only when it will facilitate better research going forward and eliminating it when it facilitates people using outdated methods because they haven't yet upgraded to current best practices.

**Comment by ChristineStawitz-NOAA** [comment]
Thanks all! This is helping firm up my thinking and translate it into more
guidance documents. I often see the misconception that maintaining
backwards compatibility is a good coding practice when in reality it can
increase bugs and makes testing more difficult. Hopefully the guidance on
Github releases we are working on will also make it clear how to provide a
legacy version via releases while not including backwards compatibility in
the new live code unless needed

I agree with SS's rationale for staying compatible with 3.24 but not
previous versions and ss3sim's rationale to not maintain compatibility.
Perhaps it is worth deciding on a per-function basis as well - it seems the
r4ss code to run a model and process output needs to be backwards
compatible but reading/writing input files from prior versions is a lower
priority.

Let me know if I can help with creating or modifying releases of either
r4ss or ss3sim to ensure a legacy stable release for prior SS versions does
continue to live on somewhere.

**Comment by iantaylor-NOAA** [comment]
Sounds good to me.  The function-specific choice of backwards compatibility makes sense, where less-often used functions could have less effort made for them. This could also be adapted anytime it seems the effort to add backwards compatibility is worth it for the potential benefits, such as the case of issue #291 which inspired a simple change that got `SS_ForeCatch` working for both 3.24 and 3.30.

The r4ss project might benefit from more formal plan for when to have a release. Looking at https://github.com/r4ss/r4ss/releases I see the recent pattern has been to issue a release in conjunction with either stock assessments being conducted local to most of us contributors ("v1.27.0" in June 2017 and "hake2019" in February 2019), or in association with a release of SS ("v1.35.0" to go with SS version 3.30.13).

r4ss would also benefit from better automated testing. The testthat test has some simple checks using SS example in versions 3.24, and three flavors of 3.30 included with the package in https://github.com/r4ss/r4ss/tree/master/inst/extdata/ and I occasionally test `SS_output` and `SS_plots` on a set of about 20 assessment models. Future releases would ideally be tested against that larger set (or a set of examples that better spans the features in SS that @k-doering-NOAA will be helping assemble in the future).

**Comment by iantaylor-NOAA** [comment]
Releases are not more formal than when this discussion first occurred, but other issues discussed here have seen progress, including focus on backwards compatibility for only a few core functions and more automated testing.
